airflow/dag_processing/processor.py

```diff
@@ -412,20 +410,20 @@ def manage_slas(self, dag: DAG, session: Session = None) -> None:
             else:
                 while next_info.logical_date < ts:
                     next_info = dag.next_dagrun_info(next_info.data_interval, restricted=False)
-
                     if next_info is None:
                         break
-                    if (ti.dag_id, ti.task_id, next_info.logical_date) in recorded_slas_query:
+                    next_run_id = DR.generate_run_id(DagRunType.SCHEDULED, next_info.logical_date)
+                    if (ti.dag_id, ti.task_id, next_run_id, ti.map_index) in recorded_sla_misses:
```

Comment:
I think this behaviour is _even_ more confusing. Let's not touch SlaMiss _at all_, and instead create a parse-time-error if someone tries to set an `sla` property on a mapped task.

(i.e. throw away most of this PR, sorry.)

My reason: the current SLA behaviour of creating the SlaMiss record against the next execution date is confusing (and likely wrong) so lets not confuse matters more by changing it to by against a future run_id that _may never exist_. 

We can come back and re-visit this once we have made SlaMiss more sensible.

Reply:
I.e. this dag should throw an error at parse time:

```python
with DAG(dag_id='test'):
    BashOperator.partial(sla=timedelta(seconds=10)).expand(bash_command=["echo true", "echo false"])
```

Reply:
>  the current SLA behaviour of creating the SlaMiss record against the next execution date is confusing (and likely wrong) so lets not confuse matters more by changing it to by against a future run_id that may never exist.

i think that the notion of "next execution date" is maybe a little misleading.  it just means next, relative to the last one examined.  so we're in `manage_slas`. we start with a task.  we look at the "last successful or skipped TI".  then we say, ok, let's look at the next run for that task -- relative to the last one that's done.  and let's see if it's failed its SLA.  (e.g. cus it is still running).  if so, let's create an SlaMiss for it.   the only time that the TI would not exist is when scheduler for whatever reason isn't even creating the TI -- e.g. because it's catchup=True, or max_active_tasks=1, or scheduler is having trouble. `next_run_id` would probably be more accurately called `curr_run_id`.  

wdyt?  for now i'll rename that variable.

> (i.e. throw away most of this PR, sorry.)

no worries if that's how it goes. we should do what we need to do, and it was a good exercise in any case.  but i'm not convinced that it's not the right change.

Reply:
though it is weird ... i don't understand why we immediately call next run here
https://github.com/apache/airflow/blob/main/airflow/dag_processing/processor.py#L414

we're already at "the next run" relative to the last completed TI -- why don't we see if _that_ run has an SLA miss?  it seems we skip to "the run _after_ the run after" the latest TI.  so that we could only get an SLA miss if airflow is 2 runs behind 🤪 

i need to actually do some live testing to understand how this actually behaves.

update: i did some live testing, and yeah, SLAs are all screwed up

here's an example:
![image](https://user-images.githubusercontent.com/15932138/158741020-7b21db86-8cbb-439c-9c66-aecbe4365e50.png)

i created a task that is sleep(300), SLA of one minute, and runs every 5 minutes.  here you can see that first TI gets no miss (expected) and the second TI does not get any miss.  the one that is created is always 2 runs ahead of the latest successful TI 🤦 

and one ahead of the one that is running.

also visible here is an odd quirk that appears to be a bug re timetables or data intervals or something in that area: the execution dates are wiggly i.e. not a value `N * timedelta + dag.start_date`.  they have seemingly random millis precision.  and as a consequence, not only are the SlaMiss records in the future, but i have found that they may _never_ correspond to TIs that actually come into existence. 🤷